File: elixir/emit_logs_direct.exs
# create connection
{:ok, connection} = AMQP.Connection.open
{:ok, channel} = AMQP.Channel.open(connection)

# get message and arguments
{severities, raw_message, _} =
  System.argv
  |> OptionParser.parse(strict: [info: :boolean, warning: :boolean, error: :boolean])
  |> case do
    {[], msg, _} -> {[info: true], msg, []}
    other -> other
  end

message =
  case raw_message do
    []    -> "Elixir log message..."
    words -> Enum.join(words, " ")
  end

# create exchange
AMQP.Exchange.declare(channel, "direct_logs", :direct)

# push a log message
for {severities, true} <- severities do
  severity = severities |> to_string
  AMQP.Basic.publish(channel, "direct_logs", severity, message)
  IO.puts " [x] Published '[#{severity}] #{message}'"
end


AMQP.Connection.close(connection)

File: ruby/emit_logs_direct.rb
#!/usr/bin/env ruby
require 'bunny'

# create connection
conn = Bunny.new
conn.start

# create channel
ch = conn.create_channel

# define a direct exchange
x = ch.direct 'direct_logs'

# get message and arguments
severity = ARGV.shift || 'info'
msg = ARGV.empty? ? 'Ruby log message...' : ARGV.join(' ')

# publish log to the exchange
x.publish(msg, routing_key: severity)
puts " [x] Sent #{msg}"

# close a connection
conn.close
